src/modules/NewPlus/NewShellExtensionContextMenu/template_item.cpp

```diff
@@ -0,0 +1,147 @@
+
+
+#include "pch.h"
+#include "template_item.h"
+#include <shellapi.h>
+#include "new_utilities.h"
+#include <cassert>
+#include <thread>
+#include <shellapi.h>
+#include <shlobj_core.h>
+
+using namespace Microsoft::WRL;
+using namespace newplus;
+
+template_item::template_item(const std::filesystem::path entry)
+{
+    path = entry;
+}
+
+std::wstring template_item::get_menu_title(const bool show_extention, const bool show_starting_digits) const
+{
+    std::wstring title = path.filename();
+
+    if (!show_starting_digits)
+    {
+        // Hide starting digits, spaces, and .
+        title = remove_starting_digits_from_filename(title);
+    }
+
+    if (show_extention || !path.has_extension())
+    {
+        return title;
+    }
+
+    std::wstring ext = path.extension();
+    title = title.substr(0, title.length() - ext.length());
+
+    return title;
+}
+
+std::wstring template_item::get_target_filename(const bool include_starting_digits) const
+{
+    std::wstring filename = path.filename();
+
+    if (!include_starting_digits)
+    {
+        // Remove starting digits, spaces, and .
+        filename = remove_starting_digits_from_filename(filename);
+    }
+
+    return filename;
+}
+
+std::wstring template_item::remove_starting_digits_from_filename(std::wstring filename) const
+{
+    filename.erase(0, min(filename.find_first_not_of(L"0123456789 ."), filename.size()));
+
+    return filename;
+}
+
+std::wstring template_item::get_explorer_icon() const
+{
+    return utilities::get_explorer_icon(path);
+}
+
+std::filesystem::path template_item::copy_object_to(const HWND window_handle, const std::filesystem::path destination) const
+{
+    // SHFILEOPSTRUCT wants the path to be terminated with two NULLs,
+    // so we use wcscpy because wcscpy_s writes non-NULLs into the rest of the buffer.
+    wchar_t double_terminated_path_from[MAX_PATH + 1] = { 0 };
+    wchar_t double_terminated_path_to[MAX_PATH + 1] = { 0 };
+#pragma warning(suppress : 4996) // don't complain about wcscpy deprecation
+    wcscpy(double_terminated_path_from, this->path.c_str());
+#pragma warning(suppress : 4996) // don't complain about wcscpy deprecation
+    wcscpy(double_terminated_path_to, destination.c_str());
```

Comment:
Wouldn't wcsncpy work here and be safer?

Reply:
Updated